Don't dismiss keyboard when tapping another text input

janicduplessis opened the pull request:
When using text inputs inside a ScrollView with `keyboardShouldPersistTaps=false` (default behavior) tapping another text input dismisses the keyboard instead of keeping it open and focusing the new text input which I think is the better and expected behavior.

See #10628 for more discussion about that. Note that this affects nothing but the behavior with text inputs unlike #10628.

cc @satya164 @MaxLap @ericvicenti

satya164 commented:
Sounds good to me!

ericvicenti commented:
Ok, this seems pretty non-controversial and desirable.

@facebook-github-bot shipit

MaxLap commented:
That's another welcome change.

How does this behave with a `<TextInput editable={false}/>`? I'm not sure what would be ideal to be honest. Just raising the question.

janicduplessis commented:
@MaxLap That's a good point, I can probably check for this props and dismiss the keyboard in this case.

janicduplessis commented:
@ericvicenti Any idea why the import failed? Let me also update this to handle disabled inputs before shipping again.

janicduplessis commented:
@MaxLap @ericvicenti Tested with editable={false} input and it does actually work already since the textinput doesn't receive the touch event.

janicduplessis commented:
Oups that broken touchables in scrollviews, probably why the import failed, fixed it :)

MaxLap commented:
I just tested this and it doesn't seem handle the `editable={false}` the way I expected.

From your commit, I edited the UIExplorerPage.js to `keyboardShouldPersistTaps = false`. Then I ran it and went to the TextInput page. There, I can tap on labels to show that keyboardShouldPersistTaps = false is dismissing the keyboard as it should. I can press other TextInputs and the keyboard stays up (very nice). However, when I tap on the editable=false that is somewhere in the middle of the page, the keyboard will remain open where I would have expected it to be dismissed.

I tested this on in the Android emulator.

janicduplessis commented:
@MaxLap Thanks for testing this, I did test on ios so the issue probably happens on android only. I guess touches are handled a bit differently for non-editable textinputs.

janicduplessis commented:
Touch target behavior was different on iOS and Android. I added a check if the view is disabled in the code that determines the touch target. That fixes the issue with disabled text inputs.

@astreet Could you sign off on this as the Android touch target stuff is more involved that originally planned. I checked which view uses `setEnabled` to make sure this won't cause any unexpected stuff. I wanted initially to just treat disabled views as not target-able but I stumbled on SwipeRefresh that uses enabled but still has target-able children. So what I did is treat the view as if it was had `pointerEvents: BOX_NONE`.

astreet commented on Libraries/Components/ScrollResponder.js:
```diff
@@ -112,6 +113,15 @@ type State = {
+function isTagInstanceOfTextInput(tag) {
+  var instance = getInstanceFromNode(tag);
+  return instance && (
+    instance.constructor.displayName === 'AndroidTextInput' ||
```
Not sure this is the best way to do instanceof in React... seems sketchy. @gaearon is there a better way?

gaearon replied:
Definitely not, `displayName` is a debugging information that may get stripped in production builds.

janicduplessis replied:
Any idea on what is the proper way to do that? That didnt feel right but was the only way I could find that did work.

mkonicek replied:
http://stackoverflow.com/questions/27824548/comparing-two-components-is-component-x-an-instance-of-component-a

janicduplessis replied:
The issue is that we have the internal instance and not the current element, it would be possible to get it by using a private field but I preferred not to. Is there a proper way to get the element from the viewTag?

Looks like we could also use uiViewClassName, which seems better than displayName. Here what the instance looks like.

![image](https://cloud.githubusercontent.com/assets/2677334/20566690/f9dcee7e-b163-11e6-9162-28ea37b53827.png)

gaearon replied:
Have you tried using `._currentElement.type`? While it's prefixed, it's safe with current reconciler implementation because this field is used as "semi public" throughout the reconciler codebase. It will break with Fiber reconciler but this doesn't matter right now because a lot of other things will break too, and we'll change them together. As long as you have a test verifying your fix works without reaching into private APIs.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/TouchTargetHelper.java:
```diff
@@ -200,16 +200,10 @@ private static boolean isTransformedTouchPointInView(
-
-    } else if (pointerEvents == PointerEvents.BOX_NONE) {
-      // This view can't be the target, but its children might
+    if (!view.isEnabled() || pointerEvents == PointerEvents.BOX_NONE) {
```
It seems like this would no longer respect PointerEvents.NONE/BOX_ONLY if the view is disabled. I think you need to explicitly handle the disabled case in each of the places we might return `view` as the target (e.g. BOX_ONLY and AUTO

janicduplessis commented:
@astreet @gaearon Thanks for your suggestions, I changed the view target code to better handle pointer events when the view is disabled. This simple changes the pointer event to avoid making this code an `if` mess. This does AUTO -> BOX_NONE and BOX_ONLY -> NONE, this makes it so that view can never be the target if disabled but it's children can still be depending on what is sepecified in pointerEvents.

Also changed to use `_currentElement.type` as per @gaearon suggestion, I also had to add caching to `requireNativeComponent` like `require` does to have referential equality when requiring the same native component multiple times.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/TouchTargetHelper.java:
```diff
@@ -200,6 +200,18 @@ private static boolean isTransformedTouchPointInView(
+    // Views that are disabled should never be the target of pointer events. However, their children
+    // can be because some views (SwipeRefreshLayout) use enabled but still have children that can
+    // be valid targets.
+    if (!view.isEnabled()) {
```
Nice!

astreet commented on Libraries/Components/ScrollResponder.js:
```diff
@@ -112,6 +114,27 @@ type State = {
 type Event = Object;
 
+if (Platform.OS === 'android') {
+  var AndroidTextInput = requireNativeComponent('AndroidTextInput', null);
```
This doesn't seem safe -- depending on the order things are loaded, we could cache a version of the native component with no view config.

Maybe we can just set nativeComponentName on the return of requireNativeComponent so that we can check `instance._currentElement.type.nativeComponentName` below? Not sure how @gaearon feels about that.

gaearon replied:
I didn't realize native components can have many versions.

I don't have a strong opinion about this—just saying you can use `._currentElement.type` and it seemed safer to me than comparing strings. You could put a static flag on it, e.g. `_currentElement.type.isReactNativeInternalTextComponent` or something like this if you're worried about multiple definitions.

janicduplessis replied:
What do you think about using `instance.viewConfig.uiViewClassName`? If that seems safe it wouldn't require any additional changes.

gaearon replied:
If it's something guaranteed to stay in production builds it should be fine.

gaearon replied:
Sounds good

janicduplessis commented:
@astreet @gaearon Updated the JS part.

astreet commented:
lgtm

@facebook-github-bot shipit
